llvm/lib/Analysis/ModuleSummaryAnalysis.cpp

```diff
@@ -99,10 +99,11 @@ extern cl::opt<bool> MemProfReportHintedSizes;
 // can only take an address of basic block located in the same function.
 // Set `RefLocalLinkageIFunc` to true if the analyzed value references a
 // local-linkage ifunc.
-static bool findRefEdges(ModuleSummaryIndex &Index, const User *CurUser,
-                         SetVector<ValueInfo, std::vector<ValueInfo>> &RefEdges,
-                         SmallPtrSet<const User *, 8> &Visited,
-                         bool &RefLocalLinkageIFunc) {
+static bool
+findRefEdges(ModuleSummaryIndex &Index, const User *CurUser,
+             SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &RefEdges,
```

Comment:
According to https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h (see the "Note" box):

```
Prefer to use ArrayRef<T> or SmallVectorImpl<T> as a parameter type.

It’s rarely appropriate to use SmallVector<T, N> as a parameter type. If an API only reads from the vector, it should use ArrayRef.
```

I see in some places you have used SmallVectorImpl for the parameter but others like here just SmallVector. These should presumably all be either SmallVectorImpl or ArrayRef.

Reply:
Thanks!  I'll go through the parameter types.  By the way, here, we have `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &`.  We can cast `SmallVector<T, 0> &` to `SmallVectorImpl<T> &`, but I don't think we can cast `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &` to `SetVector<ValueInfo, SmallVectorImpl<ValueInfo>> &`.  Note that `SmallVector` is embedded in a parent type, namely `SetVector`.